### PR TITLE
[BugFix] InsertStmt add shuffle property for non-duplicate keys

### LINKED ISSUE #6230: Data in-consistency caused by parallel of TableSink
### Steps to reproduce the behavior (Required)

The problem is hard to reproduced, here are the key conditions to reproduced it

* table with a non-duplicate keys, such as primary key or aggregate key
* replication_num > 1, such as 3
* `insert into t0 select xxx`, where `xxx` is a complex query

### Real behavior (Required)

Different copies of the same tablet replication is not the same. I.e. data in-consistency

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

+-------------------+
| current_version() |
+-------------------+
| UNKNOWN 13821230c |
+-------------------+

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #6230

## Problem Summary(Required) ：

OlapTableSink may be executed in multiply fragment instances of different machines

For non-duplicate key types, we must guarantee that the orders of the same key are exactly the same. In order to achieve this goal, we can solve this issue by performing shuffle exchange before TableSink so that the same key will be sent to the same fragment instance

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/HashDistributionDesc.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|19|19|100.00%|

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/InsertPlanTest.java [hunk 1]
```diff
@@ -566,4 +566,41 @@ public void testExplainInsert() throws Exception {
...
+            assertContains(plan, "  STREAM DATA SINK\n" +
+                    "    EXCHANGE ID: 01\n" +
+                    "    HASH_PARTITIONED: 1: pk\n" +
+                    "\n" +
+                    "  0:OlapScanNode\n" +
+                    "     TABLE: tprimary");
+        }
+        {
+            // KesType is AGG_KEYS
+            String sql = "explain insert into test_object select * from test_object";
```
Comment (kangkaisen) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/InsertPlanTest.java:596]: Would better add UT to cover the group by sql, to ensure no additional shuffle node.

**Comment by kangkaisen** [comment]
Would better create an issue for this bug.

**Comment by jaogoy** [comment]
@liuyehcf Make sure there is an issue to link and to solve.

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/InsertPlanTest.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/InsertPlanTest.java:596]: Ok, I'll make up

**Comment by Youngwb** [comment]
This can only ensure that the copy data is consistent, but does not guarantee the order of import for non-duplicate key types?

**Comment by chaoyli** [comment]
@liuyehcf Will this cause load performance degradation？ And to which extent.

**Comment by liuyehcf** [comment]
> @liuyehcf Will this cause load performance degradation？ And to which extent.

For simple insert into statement, such as `insert into t0 select * from t0`, no extra exchange node will be added.

And for some cases, such as `insert into t0 select min(k1), c2` from t0 group by c2` where c2 is not the key columns, then the extra exchange node will be added to enforce the desired distribution property

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/HashDistributionSpec.java [hunk 2]
```diff
@@ -116,9 +116,10 @@ public boolean isSatisfy(DistributionSpec spec) {
             ColocateTableIndex colocateIndex = GlobalStateMgr.getCurrentColocateIndex();
             long tableId = propertyInfo.tableId;
             // Disable use colocate/bucket join when table with empty partition
-            boolean satisfyColocate = propertyInfo.isSinglePartition() || (colocateIndex.isColocateTable(tableId) &&
-                    !colocateIndex.isGroupUnstable(colocateIndex.getGroup(tableId)) &&
-                    !propertyInfo.isEmptyPartition());
+            boolean satisfyColocate = propertyInfo.isEmptyPartition() || propertyInfo.isSinglePartition() ||
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/HashDistributionSpec.java:119]: Why change this?

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java [hunk 3]
```diff
@@ -310,4 +316,41 @@ OptExprBuilder castOutputColumnsTypeToTargetColumns(ColumnRefFactory columnRefFa
...
+
+        OlapTable table = (OlapTable) insertStmt.getTargetTable();
+
+        if (KeysType.DUP_KEYS.equals(table.getKeysType())) {
+            return new PhysicalPropertySet();
+        }
+
+        List<Column> columns = table.getFullSchema();
+        Preconditions.checkState(columns.size() == outputColumns.size(),
+                "outputColumn's size must equal with table's column size");
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java:340]: If origin table columns size is 3, but one column has default value.
we can insert only two columns, like:
```
MySQL td> desc t1;
+-------+--------+------+------+---------+-------+
| Field | Type   | Null | Key  | Default | Extra |
+-------+--------+------+------+---------+-------+
| v1    | BIGINT | Yes  | true | <null>  |       |
| v2    | BIGINT | Yes  | true | <null>  |       |
| v3    | BIGINT | Yes  | true | <null>  |       |
+-------+--------+------+------+---------+-------+
3 rows in set
Time: 0.006s
MySQL td>
MySQL td> insert into t1(v1, v2) select v1, v2 from t2;
Query OK, 2 rows affected
Time: 0.131s
MySQL td> select * from t1;
+----+----+--------+
| v1 | v2 | v3     |
+----+----+--------+
| 1  | 10 | <null> |
| 2  | 20 | <null> |
+----+----+--------+
2 rows in set
Time: 0.018s
```

**Comment by kangkaisen** [comment]
Need to check the replication number of the target table.

File: fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java [hunk 4]
```diff
@@ -310,4 +316,41 @@ OptExprBuilder castOutputColumnsTypeToTargetColumns(ColumnRefFactory columnRefFa
...
+        List<Column> keyColumns = table.getKeyColumnsByIndexId(table.getBaseIndexId());
+        List<Integer> keyColumnIds = Lists.newArrayList();
+        keyColumns.forEach(column -> {
+            int index = columns.indexOf(column);
+            Preconditions.checkState(index >= 0);
+            keyColumnIds.add(outputColumns.get(index).getId());
+        });
+
+        HashDistributionDesc desc =
+                new HashDistributionDesc(keyColumnIds, HashDistributionDesc.SourceType.SHUFFLE_JOIN);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java:351]: SHUFFLE_AGG is enough?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/HashDistributionDesc.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|19|19|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java (hunk 4 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java:351]: yes, strict order is unnecessary, I'll update

File: fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java (hunk 3 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java:340]: The following statement will collect all the specified columns and un-specified nullable columns. So the condition is always true

```java
        //3. Fill in the default value and NULL
        OptExprBuilder optExprBuilder = fillDefaultValue(logicalPlan, columnRefFactory, insertStmt, outputColumns);
```

**Review by Youngwb — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/HashDistributionSpec.java (hunk 2 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/HashDistributionSpec.java]: It has been reverted.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 20 / 20 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/HashDistributionDesc.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|19|19|100.00%|

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 23 / 23 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/HashDistributionDesc.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|22|22|100.00%|

**Comment by liuyehcf** [comment]
@Mergifyio backport branch-2.2

**Pull request merged**